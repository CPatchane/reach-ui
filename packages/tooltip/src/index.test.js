import React from "react";
import { LEAVE_TIMEOUT, MOUSE_REST_TIMEOUT } from ".";

let Tooltip;
let render, fireEvent, act, cleanup;

describe("Tooltip", () => {
  beforeEach(() => {
    ({ default: Tooltip } = require("."));
    ({ cleanup, render, fireEvent, act } = require("@testing-library/react"));

    jest.useFakeTimers();
  });

  afterEach(() => {
    cleanup();
  });

  it("shows/hides on hover", () => {
    const { baseElement, getByText } = render(
      <Tooltip label="Content">
        <button>Trigger</button>
      </Tooltip>
    );
    const trigger = getByText("Trigger");

    expect(baseElement).toMatchSnapshot();
    fireEvent.mouseOver(trigger);
    act(() => jest.advanceTimersByTime(MOUSE_REST_TIMEOUT));
    expect(baseElement).toMatchSnapshot();
    fireEvent.mouseLeave(trigger);
    act(() => jest.advanceTimersByTime(LEAVE_TIMEOUT));
    expect(baseElement).toMatchSnapshot();
  });

  it("shows/hides when trigger is activeElement", () => {
    const { baseElement, getByText } = render(
      <Tooltip label="Content">
        <button>Trigger</button>
      </Tooltip>
    );
    const trigger = getByText("Trigger");

    expect(baseElement).toMatchSnapshot();
    fireEvent.focus(trigger);
    expect(baseElement).toMatchSnapshot();
    fireEvent.blur(trigger);
    act(() => jest.advanceTimersByTime(LEAVE_TIMEOUT));
    expect(baseElement).toMatchSnapshot();
  });

  it("shows without timeout when one tooltip is already visible", () => {
    const { baseElement, getByText } = render(
      <>
        <Tooltip label="First">
          <button>First Trigger</button>
        </Tooltip>
        <Tooltip label="Second">
          <button>Second Trigger</button>
        </Tooltip>
      </>
    );
    const firstTrigger = getByText("First Trigger");
    const secondTrigger = getByText("Second Trigger");

    fireEvent.mouseOver(firstTrigger);
    act(() => jest.advanceTimersByTime(MOUSE_REST_TIMEOUT));
    expect(baseElement).toMatchSnapshot();
    fireEvent.mouseLeave(firstTrigger);
    fireEvent.mouseOver(secondTrigger);
    expect(baseElement).toMatchSnapshot();
  });

  it("hides on ESC", () => {
    const { baseElement, getByText } = render(
      <Tooltip label="Content">
        <button>Trigger</button>
      </Tooltip>
    );

    const trigger = getByText("Trigger");
    fireEvent.focus(trigger);
    act(() => jest.advanceTimersByTime(MOUSE_REST_TIMEOUT));
    expect(baseElement).toMatchSnapshot();
    fireEvent.keyDown(trigger, { key: "Escape" });
    expect(baseElement).toMatchSnapshot();
  });
});
