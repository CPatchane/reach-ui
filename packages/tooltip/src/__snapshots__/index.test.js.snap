// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tooltip hides on ESC 1`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip-1"
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip-1"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Content
    </div>
  </reach-portal>
</body>
`;

exports[`Tooltip hides on ESC 2`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`Tooltip shows without timeout when one tooltip is already visible 1`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip-1"
      data-reach-tooltip-trigger=""
    >
      First Trigger
    </button>
    <button
      data-reach-tooltip-trigger=""
    >
      Second Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip-1"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      First
    </div>
  </reach-portal>
</body>
`;

exports[`Tooltip shows without timeout when one tooltip is already visible 2`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      First Trigger
    </button>
    <button
      aria-describedby="tooltip-2"
      data-reach-tooltip-trigger=""
    >
      Second Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip-2"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Second
    </div>
  </reach-portal>
</body>
`;

exports[`Tooltip shows/hides on hover 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`Tooltip shows/hides on hover 2`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip-1"
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip-1"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Content
    </div>
  </reach-portal>
</body>
`;

exports[`Tooltip shows/hides on hover 3`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`Tooltip shows/hides when trigger is activeElement 1`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;

exports[`Tooltip shows/hides when trigger is activeElement 2`] = `
<body>
  <div>
    <button
      aria-describedby="tooltip-1"
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
  <reach-portal>
    <div
      data-reach-tooltip="true"
      id="tooltip-1"
      role="tooltip"
      style="left: 0px; top: 8px;"
    >
      Content
    </div>
  </reach-portal>
</body>
`;

exports[`Tooltip shows/hides when trigger is activeElement 3`] = `
<body>
  <div>
    <button
      data-reach-tooltip-trigger=""
    >
      Trigger
    </button>
  </div>
</body>
`;
